File: core/src/main/java/org/iumotionlabs/hefesto/feature/portinfo/service/PortInfoService.java
package org.iumotionlabs.hefesto.feature.portinfo.service;

import org.iumotionlabs.hefesto.feature.portinfo.model.*;
import org.iumotionlabs.hefesto.feature.portinfo.parser.*;

import java.io.IOException;
import java.net.ServerSocket;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * Service for orchestrating port information queries.
 */
public final class PortInfoService {

    private final PortParser parser;
    private final OperatingSystem os;
    private final ServiceRegistry serviceRegistry;

    // Common development ports
    private static final Set<Integer> DEV_PORTS = Set.of(
        3000, 3001, 4200, 5173, 5174, 8080, 8000, 8081, 8082,
        5005, 9229, 35729, 8888, 9000, 5000, 5001
    );

    public PortInfoService() {
        this.os = OperatingSystem.current();
        this.parser = createParser();
        this.serviceRegistry = new ServiceRegistry();
    }

    // For testing
    PortInfoService(PortParser parser) {
        this.os = OperatingSystem.current();
        this.parser = parser;
        this.serviceRegistry = new ServiceRegistry();
    }

    private PortParser createParser() {
        return switch (os) {
            case LINUX -> new LinuxPortParser();
            case MACOS -> new MacOsPortParser();
            case WINDOWS -> new WindowsPortParser();
            case UNKNOWN -> throw new UnsupportedOperationException(
                "Sistema operativo no soportado: " + System.getProperty("os.name")
            );
        };
    }

    /**
     * Finds bindings for a specific port.
     */
    public List<PortBinding> findByPort(int port) {
        return parser.findByPort(port, true, false);
    }

    /**
     * Finds bindings for a specific port with protocol selection.
     */
    public List<PortBinding> findByPort(int port, boolean tcp, boolean udp) {
        return parser.findByPort(port, tcp, udp);
    }

    /**
     * Finds all ports associated with a process.
     */
    public List<PortBinding> findByPid(long pid) {
        return parser.findByPid(pid);
    }

    /**
     * Finds ports in a range.
     */
    public List<PortBinding> findInRange(int from, int to, boolean listenOnly) {
        return parser.findInRange(from, to, listenOnly);
    }

    /**
     * Finds all listening ports.
     */
    public List<PortBinding> findAllListening() {
        return parser.findAllListening();
    }

    /**
     * Finds all active port bindings.
     */
    public List<PortBinding> findAll(boolean tcp, boolean udp) {
        return parser.findAll(tcp, udp);
    }

    /**
     * Finds all active port bindings (TCP and UDP).
     */
    public List<PortBinding> findAll() {
        return parser.findAll();
    }

    /**
     * Finds ports by process name (case-insensitive partial match).
     */
    public List<PortBinding> findByProcessName(String processName) {
        return parser.findByProcessName(processName);
    }

    /**
     * Finds common development ports that are in use.
     */
    public List<PortBinding> findDevPorts() {
        List<PortBinding> devBindings = new ArrayList<>();
        List<PortBinding> allListening = findAllListening();

        for (PortBinding binding : allListening) {
            if (DEV_PORTS.contains(binding.port()) || serviceRegistry.isDevelopmentPort(binding.port())) {
                devBindings.add(binding);
            }
        }

        return devBindings;
    }

    /**
     * Checks if a port is free (not in use).
     */
    public boolean isPortFree(int port) {
        // First check with quick socket test
        try (ServerSocket socket = new ServerSocket(port)) {
            socket.setReuseAddress(true);
            return true;
        } catch (IOException e) {
            return false;
        }
    }

    /**
     * Finds free ports near the given port.
     * Returns a list of available ports starting from the given port.
     */
    public List<Integer> findFreePorts(int startPort, int count) {
        List<Integer> freePorts = new ArrayList<>();
        int port = startPort;
        int maxPort = Math.min(startPort + 1000, 65535);

        while (freePorts.size() < count && port <= maxPort) {
            if (isPortFree(port)) {
                freePorts.add(port);
            }
            port++;
        }

        return freePorts;
    }

    /**
     * Finds alternative ports if the given port is in use.
     * Returns the original port if free, otherwise finds nearby alternatives.
     */
    public List<Integer> findAlternatives(int port, int count) {
        List<Integer> alternatives = new ArrayList<>();

        if (isPortFree(port)) {
            alternatives.add(port);
            count--;
        }

        // Look for alternatives near the original port
        if (count > 0) {
            alternatives.addAll(findFreePorts(port + 1, count));
        }

        return alternatives;
    }

    /**
     * Gets a network overview with all listening ports and statistics.
     */
    public NetworkOverview getNetworkOverview() {
        List<PortBinding> allBindings = findAllListening();
        List<EnrichedPortBinding> enriched = enrichBindings(allBindings);
        return NetworkOverview.from(enriched);
    }

    /**
     * Enriches a list of port bindings with service information.
     */
    public List<EnrichedPortBinding> enrichBindings(List<PortBinding> bindings) {
        List<EnrichedPortBinding> enriched = new ArrayList<>();

        for (PortBinding binding : bindings) {
            EnrichedPortBinding eb = EnrichedPortBinding.from(binding);

            // Add service info
            serviceRegistry.lookup(binding.port(), binding.protocol())
                .ifPresent(service -> {
                    enriched.add(eb.withServiceInfo(service));
                    return;
                });

            if (!enriched.contains(eb)) {
                enriched.add(eb);
            }
        }

        return enriched;
    }

    /**
     * Enriches a single port binding with service information.
     */
    public EnrichedPortBinding enrichBinding(PortBinding binding) {
        EnrichedPortBinding eb = EnrichedPortBinding.from(binding);
        return serviceRegistry.lookup(binding.port(), binding.protocol())
            .map(eb::withServiceInfo)
            .orElse(eb);
    }

    /**
     * Terminates a process.
     */
    public boolean killProcess(long pid, boolean force) {
        return parser.killProcess(pid, force);
    }

    /**
     * Returns the current operating system.
     */
    public OperatingSystem getOperatingSystem() {
        return os;
    }

    /**
     * Returns the service registry.
     */
    public ServiceRegistry getServiceRegistry() {
        return serviceRegistry;
    }

    /**
     * Enumeration of supported operating systems.
     */
    public enum OperatingSystem {
        LINUX, MACOS, WINDOWS, UNKNOWN;

        public static OperatingSystem current() {
            String osName = System.getProperty("os.name").toLowerCase();
            if (osName.contains("linux")) return LINUX;
            if (osName.contains("mac")) return MACOS;
            if (osName.contains("win")) return WINDOWS;
            return UNKNOWN;
        }
    }
}

File: hefesto-core/src/feature/portinfo/service/port_info_service.rs
use std::collections::HashSet;
use std::net::TcpListener;
use std::sync::Arc;

use hefesto_domain::portinfo::enriched_port_binding::EnrichedPortBinding;
use hefesto_domain::portinfo::network_overview::NetworkOverview;
use hefesto_domain::portinfo::port_binding::PortBinding;
use hefesto_platform::port_parser::PortParser;

use super::service_registry::ServiceRegistry;

/// Common development ports that are checked in dev-port discovery.
const DEV_PORTS: &[u16] = &[
    3000, 3001, 4200, 5173, 5174, 8080, 8000, 8081, 8082, 5005, 9229, 35729, 8888, 9000, 5000, 5001,
];

/// Service for orchestrating port information queries.
///
/// Acts as the primary facade over the platform-specific `PortParser` and the
/// `ServiceRegistry`, enriching raw port bindings with well-known service
/// metadata and providing higher-level operations such as free-port detection
/// and network overview generation.
pub struct PortInfoService {
    parser: Arc<dyn PortParser>,
    service_registry: ServiceRegistry,
}

impl PortInfoService {
    /// Creates a new `PortInfoService` with the given platform parser.
    pub fn new(parser: Arc<dyn PortParser>) -> Self {
        Self {
            parser,
            service_registry: ServiceRegistry::new(),
        }
    }

    /// Finds bindings for a specific port (TCP only by default).
    pub fn find_by_port(&self, port: u16) -> Vec<PortBinding> {
        self.parser
            .find_by_port(port, true, false)
            .unwrap_or_default()
    }

    /// Finds bindings for a specific port with protocol selection.
    pub fn find_by_port_with_protocol(&self, port: u16, tcp: bool, udp: bool) -> Vec<PortBinding> {
        self.parser.find_by_port(port, tcp, udp).unwrap_or_default()
    }

    /// Finds all ports associated with a process.
    pub fn find_by_pid(&self, pid: u32) -> Vec<PortBinding> {
        self.parser.find_by_pid(pid).unwrap_or_default()
    }

    /// Finds ports in a range.
    pub fn find_in_range(&self, from: u16, to: u16, listen_only: bool) -> Vec<PortBinding> {
        self.parser
            .find_in_range(from, to, listen_only)
            .unwrap_or_default()
    }

    /// Finds all listening ports.
    pub fn find_all_listening(&self) -> Vec<PortBinding> {
        self.parser.find_all_listening().unwrap_or_default()
    }

    /// Finds all active port bindings.
    pub fn find_all(&self, tcp: bool, udp: bool) -> Vec<PortBinding> {
        self.parser.find_all(tcp, udp).unwrap_or_default()
    }

    /// Finds all active port bindings (TCP and UDP).
    pub fn find_all_both(&self) -> Vec<PortBinding> {
        self.parser.find_all_both().unwrap_or_default()
    }

    /// Finds ports by process name (case-insensitive partial match).
    pub fn find_by_process_name(&self, process_name: &str) -> Vec<PortBinding> {
        self.parser
            .find_by_process_name(process_name)
            .unwrap_or_default()
    }

    /// Finds common development ports that are in use.
    pub fn find_dev_ports(&self) -> Vec<PortBinding> {
        let dev_set: HashSet<u16> = DEV_PORTS.iter().copied().collect();
        let all_listening = self.find_all_listening();

        all_listening
            .into_iter()
            .filter(|b| {
                dev_set.contains(&b.port) || self.service_registry.is_development_port(b.port)
            })
            .collect()
    }

    /// Checks if a port is free (not in use).
    pub fn is_port_free(&self, port: u16) -> bool {
        TcpListener::bind(("127.0.0.1", port)).is_ok()
    }

    /// Finds free ports near the given port.
    /// Returns a list of available ports starting from the given port.
    pub fn find_free_ports(&self, start_port: u16, count: usize) -> Vec<u16> {
        let mut free_ports = Vec::new();
        let max_port = start_port.saturating_add(1000);
        let mut port = start_port;

        while free_ports.len() < count && port <= max_port {
            if self.is_port_free(port) {
                free_ports.push(port);
            }
            port = match port.checked_add(1) {
                Some(p) => p,
                None => break,
            };
        }

        free_ports
    }

    /// Finds alternative ports if the given port is in use.
    /// Returns the original port if free, otherwise finds nearby alternatives.
    pub fn find_alternatives(&self, port: u16, mut count: usize) -> Vec<u16> {
        let mut alternatives = Vec::new();

        if self.is_port_free(port) {
            alternatives.push(port);
            count = count.saturating_sub(1);
        }

        if count > 0 {
            if let Some(next) = port.checked_add(1) {
                alternatives.extend(self.find_free_ports(next, count));
            }
        }

        alternatives
    }

    /// Gets a network overview with all listening ports and statistics.
    pub fn get_network_overview(&self) -> NetworkOverview {
        let all_bindings = self.find_all_listening();
        let enriched = self.enrich_bindings(&all_bindings);
        NetworkOverview::from_bindings(enriched)
    }

    /// Enriches a list of port bindings with service information.
    pub fn enrich_bindings(&self, bindings: &[PortBinding]) -> Vec<EnrichedPortBinding> {
        bindings
            .iter()
            .map(|binding| self.enrich_binding(binding))
            .collect()
    }

    /// Enriches a single port binding with service information.
    pub fn enrich_binding(&self, binding: &PortBinding) -> EnrichedPortBinding {
        let eb = EnrichedPortBinding::from_binding(binding.clone());
        match self
            .service_registry
            .lookup(binding.port, binding.protocol.as_str())
        {
            Some(service) => eb.set_service_info(service.clone()),
            None => eb,
        }
    }

    /// Terminates a process.
    pub fn kill_process(&self, pid: u32, force: bool) -> bool {
        self.parser.kill_process(pid, force).unwrap_or(false)
    }

    /// Returns a reference to the service registry.
    pub fn service_registry(&self) -> &ServiceRegistry {
        &self.service_registry
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use anyhow::Result;
    use hefesto_domain::portinfo::port_binding::{ConnectionState, Protocol};

    /// Mock parser for testing.
    struct MockPortParser {
        bindings: Vec<PortBinding>,
    }

    impl MockPortParser {
        fn new(bindings: Vec<PortBinding>) -> Self {
            Self { bindings }
        }
    }

    impl PortParser for MockPortParser {
        fn find_by_port(&self, port: u16, _tcp: bool, _udp: bool) -> Result<Vec<PortBinding>> {
            Ok(self
                .bindings
                .iter()
                .filter(|b| b.port == port)
                .cloned()
                .collect())
        }

        fn find_by_pid(&self, pid: u32) -> Result<Vec<PortBinding>> {
            Ok(self
                .bindings
                .iter()
                .filter(|b| b.pid == pid)
                .cloned()
                .collect())
        }

        fn find_in_range(
            &self,
            from: u16,
            to: u16,
            _listen_only: bool,
        ) -> Result<Vec<PortBinding>> {
            Ok(self
                .bindings
                .iter()
                .filter(|b| b.port >= from && b.port <= to)
                .cloned()
                .collect())
        }

        fn find_all_listening(&self) -> Result<Vec<PortBinding>> {
            Ok(self
                .bindings
                .iter()
                .filter(|b| b.state == ConnectionState::Listen)
                .cloned()
                .collect())
        }

        fn find_all(&self, _tcp: bool, _udp: bool) -> Result<Vec<PortBinding>> {
            Ok(self.bindings.clone())
        }

        fn find_by_process_name(&self, name: &str) -> Result<Vec<PortBinding>> {
            let lower = name.to_lowercase();
            Ok(self
                .bindings
                .iter()
                .filter(|b| b.process_name.to_lowercase().contains(&lower))
                .cloned()
                .collect())
        }

        fn kill_process(&self, _pid: u32, _force: bool) -> Result<bool> {
            Ok(true)
        }
    }

    fn make_binding(port: u16, pid: u32, name: &str) -> PortBinding {
        PortBinding::listen(port, Protocol::Tcp, pid, name)
    }

    #[test]
    fn test_find_by_port() {
        let bindings = vec![make_binding(8080, 100, "java")];
        let service = PortInfoService::new(Arc::new(MockPortParser::new(bindings)));

        let result = service.find_by_port(8080);
        assert_eq!(result.len(), 1);
        assert_eq!(result[0].process_name, "java");
    }

    #[test]
    fn test_find_by_pid() {
        let bindings = vec![
            make_binding(8080, 100, "java"),
            make_binding(8081, 100, "java"),
            make_binding(3000, 200, "node"),
        ];
        let service = PortInfoService::new(Arc::new(MockPortParser::new(bindings)));

        let result = service.find_by_pid(100);
        assert_eq!(result.len(), 2);
    }

    #[test]
    fn test_find_all_listening() {
        let bindings = vec![
            make_binding(8080, 100, "java"),
            PortBinding {
                port: 443,
                protocol: Protocol::Tcp,
                state: ConnectionState::Established,
                pid: 200,
                process_name: "curl".to_string(),
                command_line: String::new(),
                user: String::new(),
                local_address: "0.0.0.0".to_string(),
                remote_address: String::new(),
                remote_port: 0,
            },
        ];
        let service = PortInfoService::new(Arc::new(MockPortParser::new(bindings)));

        let result = service.find_all_listening();
        assert_eq!(result.len(), 1);
        assert_eq!(result[0].port, 8080);
    }

    #[test]
    fn test_enrich_binding_with_known_service() {
        let binding = make_binding(3306, 100, "mysqld");
        let service = PortInfoService::new(Arc::new(MockPortParser::new(vec![])));

        let enriched = service.enrich_binding(&binding);
        assert!(enriched.service_info.is_some());
        assert_eq!(enriched.service_info.as_ref().unwrap().name, "MySQL");
    }

    #[test]
    fn test_enrich_binding_unknown_port() {
        let binding = make_binding(12345, 100, "custom");
        let service = PortInfoService::new(Arc::new(MockPortParser::new(vec![])));

        let enriched = service.enrich_binding(&binding);
        assert!(enriched.service_info.is_none());
    }

    #[test]
    fn test_find_dev_ports() {
        let bindings = vec![
            make_binding(3000, 100, "node"),
            make_binding(3306, 200, "mysqld"),
            make_binding(8080, 300, "java"),
        ];
        let service = PortInfoService::new(Arc::new(MockPortParser::new(bindings)));

        let result = service.find_dev_ports();
        // 3000 and 8080 are dev ports; 3306 is a database port
        let ports: Vec<u16> = result.iter().map(|b| b.port).collect();
        assert!(ports.contains(&3000));
        assert!(ports.contains(&8080));
        assert!(!ports.contains(&3306));
    }

    #[test]
    fn test_find_by_process_name() {
        let bindings = vec![
            make_binding(8080, 100, "java"),
            make_binding(3000, 200, "node"),
            make_binding(9229, 200, "node"),
        ];
        let service = PortInfoService::new(Arc::new(MockPortParser::new(bindings)));

        let result = service.find_by_process_name("node");
        assert_eq!(result.len(), 2);
    }

    #[test]
    fn test_network_overview() {
        let bindings = vec![
            make_binding(8080, 100, "java"),
            make_binding(3306, 200, "mysqld"),
        ];
        let service = PortInfoService::new(Arc::new(MockPortParser::new(bindings)));

        let overview = service.get_network_overview();
        assert_eq!(overview.statistics.total_listening, 2);
    }
}
